### DISCUSSION TITLE
strongswan basic config azure 2 vms same subnet ping each other not working

### DISCUSSION THREAD

**Discussion by uisuis9** [strongswan/strongswan#1989] [General]
I have a ipsec server somewhere in AWS that I use strongswan to connect with. (I only know its tunnel details, like the tunnelid, rightsubnet, right ip etc)

>

Issue is with my setup on the other end that runs strongswan.

My strongswan system is in azure.

In azure I have two vms both belonging to same subnet: 192.168.0.0/28

vm1 192.168.0.4 and has public ip, 
vm2 192.168.0.5 (has no public ip)

Im doing ipsec start from vm1.

The ipsec.conf has leftsubnet=192.168.0.0/28.

Tunnel comes up and I can access any app running in vm1.

### PROBLEM:

> If tunnel is up - FROM VM1 i CANNOT ping VM2

> If tunnel is down - FROM VM1 i CAN ping VM2

Seems basic but I tried a lot and have at my wits end. If you can spot the error please let me know.

### Kernel table when ipsec up tunnel is connected
```
Kernel IP routing table
Destination     Gateway         Genmask         Flags   MSS Window  irtt Iface
default         192.168.0.1     0.0.0.0         UG        0 0          0 eth0
168.63.129.16   192.168.0.1     255.255.255.255 UGH       0 0          0 eth0
169.254.169.254 192.168.0.1     255.255.255.255 UGH       0 0          0 eth0
192.168.0.0     0.0.0.0         255.255.255.240 U         0 0          0 eth0
```
====
The vms are ubuntu on azure and i have enabled interface forwarding in azure vm1 and also set various sysctl options in vm1 (where strongswan is running)
```
net.ipv4.ip_forward = 1
net.ipv4.conf.all.send_redirects = 0
net.ipv4.conf.default.send_redirects = 0
net.ipv4.conf.eth0.send_redirects = 0
```

**Comment by tobiasbrunner** [comment]
You haven't provided much information on what you actually configured (see #196). But you might have to configure [bypass policies](https://docs.strongswan.org/docs/5.9/config/quickstart.html#_passthroughbypass_policies) for that local subnet.

**Reply by uisuis9** [reply]
thank you. my entire config is below:
```
conn tun
        keyingtries=1
        installpolicy=yes
        keyexchange=ikev2
        dpdaction=restart
        dpddelay=30s
        dpdtimeout=150s
        reauth=no
        fragmentation=yes
        forceencaps=yes
        mobike=no
        type=tunnel
        left=%any
        leftid=*****@****.com
        leftauth=psk
        leftsubnet=192.168.0.0/28
        right=**.**.145.74
        rightauth=psk
        rightsubnet=0.0.0.0/0
        closeaction=restart
        auto=add
```
I built it using strongswan 5.9.11 using:
./configure --prefix=/usr --sysconfdir=/etc --enable-openssl --enable-curl --enable-farp
I can recompile with the bypass flag. 

Is there a wiki to use bypass lan in ipsec.conf file format to bypass subnets? Ill give it a try shortly.
(I found this https://wiki.strongswan.org/projects/strongswan/wiki/UsableExamples#Passthrough-policy)

Also, my branch is very simple. just two vms. One running ipsec up tunnel 192.168.0.4/28, and the ohter running a python webserver 192.168.0.5/28 (same subnets)

**Reply by uisuis9** [reply]
Bypass helped me ping the next vm. i dint add anything to ipsec.conf, just enabled plugin via  ./configure
however how can i access the python server from the other side via VPN?

Windows server has vpn client:
belongs to subnet: 192.168.1.4/28,
Gets 10.10.1.130 from vpn server when connected
From windows server i can do  curl http://192.168.0.4:8080 (VM1) works.
i want to do  curl http://192.168.0.5:8080 (VM2) also which does not work

so in addition to bypass lan between vm1, vm2 should i do anything else to reach other VMs in the same subnet as VM1?

**Reply by tobiasbrunner** [reply]
You might want to read about [forwarding traffic](https://docs.strongswan.org/docs/5.9/howtos/forwarding.html) and consider how traffic from and to that other host looks like and is processed by your VM1 and VM2.

**Reply by uisuis9** [reply]
Thanks will do that.
Just in case the kernel tables at 192.168.0.4 and .5 are

192.168.0.4
```
Kernel IP routing table
Destination     Gateway         Genmask         Flags   MSS Window  irtt Iface
default         192.168.0.1     0.0.0.0         UG        0 0          0 eth0
168.63.129.16   192.168.0.1     255.255.255.255 UGH       0 0          0 eth0
169.254.169.254 192.168.0.1     255.255.255.255 UGH       0 0          0 eth0
192.168.0.0     0.0.0.0         255.255.255.240 U         0 0          0 eth0
```

192.168.0.5
```
Kernel IP routing table
Destination     Gateway         Genmask         Flags   MSS Window  irtt Iface
default         _gateway        0.0.0.0         UG        0 0          0 eth0
168.63.129.16   _gateway        255.255.255.255 UGH       0 0          0 eth0
169.254.169.254 _gateway        255.255.255.255 UGH       0 0          0 eth0
192.168.0.0     0.0.0.0         255.255.255.240 U         0 0          0 eth0
```

**Reply by uisuis9** [reply]
Can you please nudge me a bit more in the right direction.
Im assuming vpn gateway in my case 192.168.0.4 (where ipsec is established)? Also doc says something about roadwarriors which I dont know anything about. request a bit more direction please.

a) I enabled in VM1 azure's portal enable forwarding (so src/dst checks dont happen)
b) I have also have sysctl options set for ipv4 forwarding=1 in VM1
c) in VM2 i have added a route to point 0.0.0.0 to VM1 and 10.10.1.1 to VM1 too
d) 10.10.1.1 is my windows ip address after connecting to vpn.

what else am i missing? would you know?

```
netstat -r on vm2
Kernel IP routing table
Destination     Gateway         Genmask         Flags   MSS Window  irtt Iface
default         192.168.0.4 0.0.0.0         UG        0 0          0 eth0
10.10.1.1       192.168.0.4 255.255.255.255 UGH       0 0          0 eth0
168.63.129.16   192.168.0.1     255.255.255.255 UGH       0 0          0 eth0
169.254.169.254 192.168.0.1     255.255.255.255 UGH       0 0          0 eth0
192.168.0.0     0.0.0.0         255.255.255.240 U         0 0          0 eth0
```

### tcpdump on vm1
```
sudo tcpdump src 10.10.1.1
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes

17:46:45.602307 IP 10.10.1.1 > 192.168.0.5: ICMP echo request, id 1, seq 2690, length 40
17:46:45.602379 IP 10.10.1.1 > 192.168.0.5: ICMP echo request, id 1, seq 2690, length 40
```

### tcpdump on vm2
```
sudo tcpdump | grep ICMP
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
```

**Reply by uisuis9** [reply]
Also the documentation is great but it is very difficult to follow:

As an example? ## "The virtual IPs are from the subnet behind the gateway"
what does this line mean? they are the same subnet as the gateway?
Should i enable farp for this to work for me?

**Reply by tobiasbrunner** [reply]
> Can you please nudge me a bit more in the right direction.

You're on the right path.

>  Also doc says something about roadwarriors which I dont know anything about. 

See [the introduction](https://docs.strongswan.org/docs/5.9/howtos/introduction.html) (in your scenario you don't have a classic roadwarrior, but the problem with forwarding the traffic from hosts at the other end is similar).

> I enabled in VM1 azure's portal enable forwarding (so src/dst checks dont happen)

Wouldn't that be necessary for both? Might be the reason why you don't capture any packets on VM2. (But I don't know what exactly Azure might be checking.)

> in VM2 i have added a route to point 0.0.0.0 to VM1 and 10.10.1.1 to VM1 too

If you add a default route via VM1, then you don't need one for 10.10.1.1. It really depends if you want to route everything via VPN from VM2 (you do from VM1 with the negotiated remote traffic selector of `0.0.0.0/0`, which is why that bypass policy for the local subnet was necessary).


> "The virtual IPs are from the subnet behind the gateway"
> what does this line mean? they are the same subnet as the gateway?
> Should i enable farp for this to work for me?

No, you don't need the _farp_ plugin. Virtual IPs are IP addresses that a VPN gateway assigns to clients (roadwarriors) from a pool (see [the docs](https://docs.strongswan.org/docs/5.9/features/vip.html)). That's not the case here and the IP address (`10.10.1.1`) is not part of your local subnet (`192.168.0.0/28`).
